plugins/auth-backend-module-oidc-proxy-provider/src/helpers.ts

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { LoggerService } from '@backstage/backend-plugin-api';
+import { AuthenticationError } from '@backstage/errors';
+import { JWK, KeyLike, importJWK, jwtVerify } from 'jose';
+import fetch from 'node-fetch';
+import { IDTokenInfo } from './types';
+
+export function createTokenValidator(
+  logger: LoggerService,
+  iss: string,
+  aud: string,
+  providedJwks?: (KeyLike | Uint8Array)[],
+): (token: string) => Promise<IDTokenInfo> {
+  let jwksUri: string = '';
+
+  return async function tokenValidator(token) {
+    // Perform discovery once to avoid two round trips on every authentication request.
+    if (jwksUri === '' && providedJwks === undefined) {
+      jwksUri = await discoverJwksUri(iss);
+    }
+
+    // Verify the token was signed by the issuer.  Performs one round trip to
+    // the jwks uri every authentication request to fetch the current key set
+    // from the issuer.  May be optimized in the future, but this happens only
+    // at sign-in, not for every request to the backend.
+    //
+    // Refer to https://github.com/panva/jose/blob/v5.4.0/docs/functions/jwks_remote.createRemoteJWKSet.md#returns-1
+    const jwks = providedJwks ?? (await loadJwksUri(logger, jwksUri));
```

Comment:
> I'm still not entirely clear on why you didn't just use createRemoteJWKSet - that's what we do (passing the key set as the second argument to jwtVerify), and it keeps itself in sync perfectly as far as we know in the background, refreshing itself from the remote when needed etc.

The reason is the identity provider Holos uses is ZITADEL.  ZITADEL sometimes returns an empty jwks, `[]`, which caused `createRemoteJWKSet` to fail without grace, which in turn caused Backstage to fail with a mysterious user-facing error.

So, this is handles that edge case.

Refer to https://github.com/zitadel/zitadel/issues/6949#issuecomment-2035030448

Copied from https://github.com/backstage/backstage/pull/25105#discussion_r1660025639

Reply:
@freben I'm going to take a deeper look into using createRemoteJWKSet to align with the rest of Backstage.  Marking this as draft until then.

Reply:
@freben I see how it's called in `backstage-defaults`: [this.#keyStore = createRemoteJWKSet(endpoint);](https://github.com/jeffmccune/backstage/blob/c63678419fa4c5ab4337f1abaa2567322729af21/packages/backend-defaults/src/entrypoints/auth/JwksClient.ts#L71)

May I `import { JwksClient } from '@backstage/backend-defaults/src/entrypoints/auth/JwksClient';` here, or does that couple the two too tightly?  Alternatively, could/should I copy the JwksClient.ts file into this plugin?

It's unclear to me how best to reuse / align with the existing JwksClient code in backstage.  Please advise, otherwise I'll proceed with copying the file to reduce coupling.  It's also easy enough to change the import.